### ISSUE TITLE
Need help to save the after abel transformed image as a 16 bit tiff file

### ISSUE THREAD

**Issue by kfxnightmare2** [PyAbel/PyAbel#303]
I tried few ways to save the tif file but all those saving it as a 32bit?

from PIL import Image
pic = Image.fromarray(recon)
pic.save('test-pillow.tif','TIFF')

but the dtype is float32, I tried to convert it to 16 bit but some forums say 16 bit conversion is not accurate in pillow 

Also,
is there a way to load the raw_data in 16-bit

raw_data = plt.imread(filename).astype('float64')  <------- instead of float64

**Comment by MikhailRyazanov** [comment]
It's not very clear what exactly you want to do, but saving a [16-bit grayscale](https://pillow.readthedocs.io/en/stable/handbook/concepts.html#modes) TIFF seems to work like this:
```python
pic = Image.fromarray(recon.astype(np.uint16), mode='I;16')
pic.save('test-pillow.tif')
```
You might need to normalize your array to the 16-bit unsigned-integer range (0 to 65 535), and perhaps do proper rounding, before applying `.astype()`.

Regarding loading 16-bit data, as far as I understand, Pillow can load 16-bit grayscale TIFFs, at least those saved as shown above. So
```python
pic = Image.open('test-pillow.tif')
raw_data = np.asarray(pic)
```
creates a NumPy array with `dtype=uint16`, that is 16-bit unsigned integers. However, Abel transform is not designed for integers, neither are other PyAbel tools, so using this loaded raw data without converting it to a floating-point format looks like a very strange idea. Do you really need that and why?

**Comment by MikhailRyazanov** [comment]
Another general remark about using image formats for saving/loading raw data: beware of [gamma correction](https://en.wikipedia.org/wiki/Gamma_correction). Most image formats use or assume nonlinear intensity encoding by default, but all data analysis must be done with intensity in a linear scale. So you must either make sure that linear encoding (γ = 1) is used everywhere in the data path, or do appropriate intensity transforms yourself.

**Comment by stggh** [comment]
An excellent point! This comment should not be lost, perhaps, it could be worked into the `PyAbel` docs?

**Comment by MikhailRyazanov** [comment]
> An excellent point!

Which one? About gamma correction?

**Comment by stggh** [comment]
Yes, the gamma correction.

**Comment by MikhailRyazanov** [comment]
OK. Gamma correction is a difficult matter. According to [TIFF specifications](https://www.adobe.io/content/dam/udp/en/open/standards/tiff/TIFF6.pdf) (p. 73):
_“Be aware that the PhotometricInterpretation value of 0 or 1 (grayscale) implies linear data because no gamma is specified. The PhotometricInterpretation value of 2 (RGB data) specifies the NTSC gamma of 2.2 as a default. If data is written as something other than the default, then a GrayResponseCurve field or a TransferFunction field must be present to define the deviation. For grayscale data, be sure that the densities in the GrayResponseCurve are consistent with the PhotometricInterpretation field and the HalftoneHints field.”_
So for _grayscale_ images TIFF _by default_ implies _linear_ intensities (although the same document on p. 22 says that for grayscale images only 4 or 8 bits per pixel are allowed, and 16-bit grayscale images are not even mentioned).
However, for example, ImageMagick [assumes non-linear grayscale](https://imagemagick.org/script/color-management.php) by default, apparently even for TIFF images. I don't know about other programs and whether they read/write these “GrayResponseCurve”/“TransferFunction” correctly when they are present...

By the way, we have `Xenon_ATI_VMI_800_nm_649x519.tif` among examples. I'm wondering whether @DanHickstein can say anything about its origin and properties. The comments within the file say “`ImageJ=1.48v`”, so it was probably saved by the [ImageJ](https://en.wikipedia.org/wiki/ImageJ) Java program, but how did it get there and how that program treats intensities?

**Comment by kfxnightmare2** [comment]
Thank you very much for your valuable answers and concern in this matter.

@MikhailRyazanov , i want to export the abel transformed image as a tiff and load that tiff into a program made by me in LabView for further analysis. As you suggested i will try those and meanwhile keep an eye on the gamma correction.  Thanks again for your support on this

**Comment by MikhailRyazanov** [comment]
> ... program made by me in LabView for further analysis.

Then I really recommend to transfer the data as a floating-point array instead of a TIFF image to avoid unnecessary intensity scaling and rounding and possible nonlinearities. I did not work with images in LabView, but if I understand correctly, it also supports floating-point images. Even if for some reason you need an integer format, it's better to do the conversion inside LabView, as is it more likely to be implemented correctly.

**Comment by DanHickstein** [comment]
Woah, I honestly had no idea about this gamma correction business before. It could really cause havoc with the analysis of images.

> I'm wondering whether @DanHickstein can say anything about its origin and properties. 

That image was collected almost a decade ago at this point, so I might not have all the details :). But yes, I probably used imageJ to background-subtract that image, so that's why the file says that it was created with ImageJ. I'm now curious if the camera used any gamma encoding. I used an AVT Pike camera [[manual link](https://www.1stvision.com/cameras/AVT/dataman/Pike_TechMan_V4.1.0_en.pdf)] to collect those images. I didn't enable any gamma encoding explicitly, but it's possible that it was turned on by default. The camera does support gamma=0.7 or gamma=0.45. It's not clear from the manual if gamma is on by default.

**Comment by MikhailRyazanov** [comment]
@DanHickstein,
> I honestly had no idea about this gamma correction business before

Then we indeed should add a couple of words about it. :–) Probably to “Conventions” in README? Something like

* **Intensity:** Absolute pixel intensities can be arbitrary (withing the floating-point range), but the intensity scale must be linear. Keep in mind that cameras and common image formats can use [gamma correction](https://en.wikipedia.org/wiki/Gamma_correction) and thus provide data with nonlinear intensity encoding. When working with such data, make sure to apply appropriate intensity transformations before the analysis. Most PyAbel methods also assume intensities to be floating-point numbers, and when applied to integer types, can return inappropriately rounded results. The `abel.Transform` class recasts the input image to `float64` by default, but if you wish to call the transform methods directly or use other tools, you might need to perform the conversion yourself (as `IM.astype(float)`, for example).

>  It's not clear from the manual if gamma is on by default.

Yes. Hopefully, it is not.
I was using different cameras (from Pixelink and Point Grey) and in my acquisition programs was always explicitly disabling the gamma correction — probably just to be sure, I don't remember whether it was enabled by default, and the only Point Grey camera that I currently have (a different model) seems to have the gamma correction disabled at power-on.

**Comment by stggh** [comment]
Collecting an image by co-adding camera images is also a bad idea in terms of accumulating camera noise. It is better to sub-pixel centroid, and count each event, where possible.

Your important addition to the docs looks good to me.

**Comment by MikhailRyazanov** [comment]
> It is better to sub-pixel centroid, and count each event, where possible.

When possible. :–) I remember, Dan said that with their femtosecond lasers with ≫1 kHz repetition rates, it's really possible to get a full, analyzable image from a single camera frame, so centroiding would not work for such high signal levels. Plus, many people really work with continuous distributions (plasma densities and so on).
But yes, with event counting, the gamma correction of the camera is less important (although it can still affect subpixel accuracy, and thus I was disabling it).

**Comment by MikhailRyazanov** [comment]
> Your important addition to the docs looks good to me.

Thanks! I've created #305, please review/comment there...

**Comment by stggh** [comment]
Yes, "when". My experiment is limited to 10 Hz. Oh to have a 1kHz repetition rate!  I think in that case, centroiding using the computer processor in the camera, in parallel with the image collection, writing only the (row, column) data point positions to file.  ;-)

**Comment by MikhailRyazanov** [comment]
I was doing subpixel centroiding at 100 Hz for 1200×1200 images on a relatively cheap computer about 5 years ago. The main problems are a sufficiently fast camera (although now you can get up to 522 Hz at 720×540 (BFS-U3-04S2M-CS) quite cheaply) and the laser, of course...

**Comment by stggh** [comment]
You are a pretty nifty scientist and programmer!  I meant using the camera CPU to perform the centroiding, without the need to transfer any images. Unfortunately, I don't know of any open source cameras.

**Comment by MikhailRyazanov** [comment]
Yes, since centroiding is in principle well localized, it could be done in parallel by something like a simple neural network built in the sensor, without transferring the whole image and thus much faster. But I also don't know any such cameras yet...

**Comment by DanHickstein** [comment]
> It is better to sub-pixel centroid, and count each event, where possible.

Yeah, this is what all of the cool scientists do :). As Mikhail mentions, I was running at 1 kHz and had a huge amount of signal. I tried using a really fast CMOS camera with a small region-of-interest, which (in some intensity regimes) could deliver me frames with ~20 electron hits. I could almost make this work, but I had trouble getting the computer to process the images fast enough. I eventually gave up and just averaged images. I think that I had bigger problems with my experiment than sensor noise. For example, my electrostatic lens was not particularly well built, and probably introduced some significant distortions to the image that made analysis more difficult. I think that I would have seriously benefited from the circularization algorithms that are now implemented in PyAbel.

There are some cameras on the market the offer some user-flexibility with on-board FPGA-based image processing. For example, these cameras look interesting: http://www.i-cube-tech.co.jp/en/products/sc_plus/
